R/utils.R

```diff
@@ -259,85 +259,560 @@ rowcol_to_dim <- function(row, col) {
+  }
 
-  # wb_dims(rows, cols)
-  if (length(args) >= 2 && !exp_name) {
-    rows <- 1L
-    cols <- 2L
+  x_present <- "x" %in% nams
+  cond_acceptable_lengt_1 <- x_present || !is.null(args$from_row) || !is.null(args$from_col)
 
-    # wb_dims(rows = rows, cols = cols)
-    sel <- pmatch(nams, c("rows", "cols"))
-    valid <- length(sel[!is.na(sel)])
-    if (valid == 2) {
-      rows <- sel[rows]
-      cols <- sel[cols]
-    } else if (valid == 1) {
-      stop("found only one cols/rows argument")
+  if (lengt == 1 && !cond_acceptable_lengt_1) {
+    # Providing a single argument acceptable is only  `x`
+    sentence_unnamed <- ifelse(all_args_unnamed, "unnamed ", " ")
+    stop(
+      "Supplying a single", sentence_unnamed, "argument to `wb_dims()` is not supported.",
+      "\n",
+      "use any of `x`, `from_row` `from_col`. You can also use `rows` and `cols`, You can also use `dims = NULL`"
+    )
+  }
+  cnam_null <- is.null(args$col_names)
+  rnam_null <- is.null(args$row_names)
+
+
+  if (!x_present) {
+    if (!cnam_null || !rnam_null) {
+      stop("`row_names`, and `col_names` should only be used if `x` is present.")
+    }
+  }
+  rows_arg <- args$rows
+  #
+  x <- args$x
+  x_has_named_dims <- inherits(x, "data.frame") | inherits(x, "matrix")
+  x_has_colnames <- !is.null(colnames(x))
+  if (x_has_colnames && !is.null(rows_arg)) {
+    # Not checking whether it's a row name, not supported.
+    is_rows_a_colname <- rows_arg %in% colnames(x)
+
+    if (any(is_rows_a_colname)) {
+      stop("`rows` is the incorrect argument. Use `cols` instead. Subsetting rows by name is not supported.")
     }
+  }
+  if (is.character(rows_arg)) {
+    warning("It's preferable to specify integers indices for `rows`", "See `col2int(rows)` to find the correct index.")
+  }
 
-    rows <- as.integer(args[[rows]])
-    cols <- col2int(args[[cols]])
+  rows_arg <- col2int(rows_arg)
+  cols_arg <- args$cols
+  x <- args$x
+  # rows_and_cols_present <- all(c("rows", "cols") %in% nams)
 
-  } else {
 
-    if (cnam_null) col_names <- exp_name
-    if (rnam_null) row_names <- FALSE
+  # Find column location id if `cols` is named.
+  if (x_has_colnames && !is.null(cols_arg)) {
+    is_cols_a_colname <- cols_arg %in% colnames(x)
 
-    assert_class(col_names, "logical")
-    assert_class(row_names, "logical")
+    if (any(is_cols_a_colname)) {
+      if (length(is_cols_a_colname) != 1) {
+        stop(
+          "Supplying multiple column names is not supported by the `wb_dims()` helper, use the `cols`  arguments instead.",
+          "\n Use a single `cols` at a time with `wb_dims()`"
+        )
+      }
+      # message("Transforming col name = '", cols_arg, "' to `cols = ", which(colnames(x) == cols_arg), "`")
+      cols_arg <- which(colnames(x) == cols_arg)
 
-    # wb_dims(data.frame())
-    x <- as.data.frame(x)
-    rows <- seq_len(nrow(x) + col_names)
-    cols <- seq_len(ncol(x) + row_names)
+    }
+  }
+
+  if (!is.null(rows_arg)) {
```

Comment:
can `row_arg` be moved in here? I do not like having code spread in and out of functions. Reading the if condition below this looks like something is missing. If this can only be integer, why do we need the assert afterwards?

Reply:
Related to 8be9dd8697b7fd13a24be572d6313c02117a6b18. In some tests or examples, characters were used to specify rows.. At some point later, it could be changed, but until rows/start rows can be characters, we still need `col2int()` I think..